pkg/kubelet/kubelet.go

```diff
@@ -1936,6 +1942,15 @@ func (kl *Kubelet) syncTerminatingPod(_ context.Context, pod *v1.Pod, podStatus
 		return fmt.Errorf("detected running containers after a successful KillPod, CRI violation: %v", runningContainers)
 	}
 
+	// NOTE: resources must be unprepared AFTER all containers have stopped
+	// and BEFORE the pod status is changed on the API server
+	// to avoid race conditions with the resource deallocation code in kubernetes core.
+	if utilfeature.DefaultFeatureGate.Enabled(features.DynamicResourceAllocation) {
```

Comment:
Thanks for moving this -- this is better to put here as part of `syncTerminatingPod`. However, I still think there is a possibility of a race here based on my understanding -- take the case of eviction:

1. Pod will be evicted with `killPod` (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/eviction/eviction_manager.go#L584-L591)
2. killPod sends a `SyncPodKill` to the pod worker - https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/pod_workers.go#L1305-L1314
3. Pod worker will start to tear down the pod by sending `TerminatingPodWork` (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/pod_workers.go#L682) - https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/pod_workers.go#L755-L759
4. Pod worker calls `syncTerminatingPod` (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/pod_workers.go#L945-L947) passing the status update function from eviction manager (which will update the status)
5. The status is updated to the status manager above on line 1885 - `kl.statusManager.SetPodStatus(pod, apiPodStatus)`
6. The status manager loops run in a seperate go-routine and is allowed to update the API status if the all containers have terminated (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/status/status_manager.go#L918-L933)
7. We only after call  `UnprepareResources`

So in summary, based on my understanding I think there is still a race here between calling `UnprepareResources` and the status being sent. If we want ensure that `UnprepareResources` is done before the status is sent, we need to update the logic in status_manager to ensure that all containers have terminated as well as the resources have been cleaned up by `UnprepareResources` (i.e. this logic https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/status/status_manager.go#L918-L933)

Reply:
Thanks for the detailed response and suggestions @bobbypage.

Would you be OK with something as simple as the following to unblock this by tomorrow:
```
// PodCouldHaveRunningContainers returns true if the pod with the given UID could still have running
// containers. This returns false if the pod has not yet been started or the pod is unknown.
func (kl *Kubelet) PodCouldHaveRunningContainers(pod *v1.Pod) bool {
-	return kl.podWorkers.CouldHaveRunningContainers(pod.UID)
+	if kl.podWorkers.CouldHaveRunningContainers(pod.UID) {
+		return true
+	}
+
+	// Detailed comment about why this is here
+	if utilfeature.DefaultFeatureGate.Enabled(features.DynamicResourceAllocation) {
+		if kl.containerManager.HasPreparedResources(pod) {
+			return true
+		}
+	}
+
+	return false
}
```

With a follow-up issue to figure out how to do this "the right way" after the release.

This would reduce churn in the existing code, base, but temporarily conflate the concept of "running containers" and "prepared resource claims" in the `PodDeletionSafetyProvider` until we built out the proper abstraction / plumbing to represent this new concept.

I think the thing that makes this hard is that there is already a function in the `PodDeletionSafetyProvider` interface called `PodResourcesAreReclaimed()` which sounds like it would be the perfect place for our new logic -- except that it is only used to prevent a pod from being deleted -- not entering the terminal state.

Maybe, we want a new function in this interface -- or maybe we just want to rename `PodCouldHaveRunningContainers()` to something that captures the fact that both of these checks are necessary in here. At least in the current state of the code this would suffice, since this function it is only ever used to perform the exact check we are wanting to also gate the "unpreparing" of our resources on.

Reply:
Alternatively, we just leave this out for now, merge the code as-is with this known race condition, and follow-up on it after the release (I can create another issue for this).

Given that all of the code is protected by the feature-gate, it is very low-risk to any users that aren't opting in to the feature (that could still be buggy in other areas given that its still in `alpha`).

The important thing being that we have identified this race during the review so we can follow up quickly after.

Reply:
Yet another alternative: https://github.com/pohly/kubernetes/pull/46
The PR may look like a big change, but it's pretty simple. Most of the changes are done to add new interface to all structures in a call stack from the status manager to the DRA manager.

All DRA e2e tests pass with the change:
```
Ran 18 of 6811 Specs in 271.840 seconds
SUCCESS! -- 18 Passed | 0 Failed | 0 Pending | 6793 Skipped
--- PASS: TestE2E (271.98s)
PASS
ok  	k8s.io/kubernetes/test/e2e	272.097s
```

Reply:
Thanks @klueska and @bart0sh .

The suggestion @klueska of adding this check into `PodCouldHaveRunningContainers` -- that would work as well, it looks like it's only used for gating transitioning to terminal phase which is exactly the scenario we need here. So checking there if pod is referencing a claim would work. 

I agree we would need to followup and refactor it a bit, perhaps by renaming it to `PodCanTransitionToTerminalPhase` or something like that, since it in that case we would extend it's logic to more than just running containers. The solution that @bart0sh proposed in https://github.com/pohly/kubernetes/pull/46 would work as well, but it requires a bit more plumbing, so maybe adding this to `PodCouldHaveRunningContainers` is the most straightforward solution for now.

Overall, I agree -- I don't think see this as a alpha blocker (the changes are feature gated, so they shouldn't cause any regressions to folks not using this feature), but it's definitely something we need to track and fix.

Reply:
> so maybe adding this to PodCouldHaveRunningContainers is the most straightforward solution for now.

So let's do that, right?

@bart0sh : can you create an issue for the future cleanup work and turn the draft patch from above into a commit that I can add?

Reply:
@pohly actually given @bobbypage's comment here:
> Definitely agree -- I don't think see this as an alpha blocker (the changes are feature gated, so they shouldn't cause any regressions to folks not using this feature), but it's definitely something we need to track and fix.

I'm inclined to hold off adding anything new at this stage. That way we can think carefully about how we want to add this the right way, complete with comprehensive tests.

Reply:
I had added a comment here that we should just hold off adding anything at this stage that I just deleted.

I hadn't seen the ongoing slack conversation discussing the change here:
https://kubernetes.slack.com/archives/C0BP8PW9G/p1668124328460089?thread_ts=1668065189.545059&cid=C0BP8PW9G

Looks like most of the details have been worked out, so let's move forward with the proposed patch, with an issue to clean it up and add comprehensive tests later.

Reply:
> @bart0sh : can you create an issue for the future cleanup work and turn the draft patch from above into a commit that I can add?

Done: https://github.com/kubernetes/kubernetes/issues/113851

Reply:
I have included the fix for this from https://github.com/pohly/kubernetes/pull/46

Reply:
Thanks, lgtm on the fix. Can we please create a tracking issue to track the TODO mention there so it can be addressed as part of beta:

```
// NOTE: This is a temporary solution. This call is here to avoid changing
// status manager and its tests.
// TODO: extend PodDeletionSafetyProvider interface and implement it
// in a separate Kubelet method.
```

Reply:
I just saw this while reviewing recent changes in the pod worker and this really needs to be fixed.  If dynamic resources need special status information to coordinate with other actors, that needs to be provided as part of the status reporting of the Kubelet.  I would probably have blocked this from merging if I'd caught it, just because correctness of the kubelet is so concerning in this area.

What's the latest on this?

Reply:
@smarterclayton Here is the issue about it https://github.com/kubernetes/kubernetes/issues/113851. It's in progress.

Reply:
@smarterclayton @bobbypage  This should be fixed by https://github.com/kubernetes/kubernetes/pull/116263, PTAL